Unit tests for MiCS

zarzen opened the pull request:
In response to the ask from https://github.com/microsoft/DeepSpeed/pull/2964#issuecomment-1832161865, I added three more unit tests related to MiCS. 

There are two knowledge issues:
- Testing on Torch 2.1.0 triggers `_IllegalWorker` in coalesced all gather. I made changes to ignore this condition. and Currently, I don't know the reason. 
- The MiCS implementation is not working with offloading, so the failure in `TestZeroPartialOffloadConfigSweep` is expected.

tjruwase commented:
@zarzen, thanks!

mrwyattii commented on tests/unit/runtime/zero/test_zero.py:
```diff
@@ -85,18 +85,25 @@ def test(self, zero_stage):
 
 
 # testing the fix https://github.com/microsoft/DeepSpeed/pull/1227
+@pytest.mark.parametrize("zero_stage", [3, "mics"])
```
```suggestion
@pytest.mark.parametrize("mics_enabled", [True, False])
```

mrwyattii commented on tests/unit/runtime/zero/test_zero.py:
```diff
@@ -85,18 +85,25 @@ def test(self, zero_stage):
     world_size = 1
 
-    def test(self, zero_stage=3):
+    def test(self, zero_stage):
```
```suggestion
    def test(self, mics_enabled, zero_stage=3):
```

mrwyattii commented on tests/unit/runtime/zero/test_zero.py:
```diff
@@ -85,18 +85,25 @@ def test(self, zero_stage):
-    def test(self, zero_stage=3):
+    def test(self, zero_stage):
         # force all params to be partitioned by forcing threshold=0
+        mics_shard_size = -1
+        if isinstance(zero_stage, str):
+            zero_stage = 3
+            mics_shard_size = self.world_size
+
```
```suggestion
        mics_shard_size = -1
        if mics_enabled:
            mics_shard_size = self.world_size
```

mrwyattii reviewed (changes requested):
I think it will be more clear if we parameterize a `mics_enabled` variable rather than use `zero_stage` to switch it on/off. Please update the other test with similar changes. Thanks!

zarzen commented:
@mrwyattii thanks for suggestions, I updated the implementation accordingly. Besides, as the mics implementation currently is not compatible with offloading, I removed the unittest for `test_zero_offloadpp.py`
